diff --git a/praxis/base_layer.py b/praxis/base_layer.py
index cc5f24c..84502d4 100644
--- a/praxis/base_layer.py
+++ b/praxis/base_layer.py
@@ -246,7 +246,7 @@ def to_partition_spec(
               f's_i: {s_i}, mesh_axis_names: {mesh_axis_names}')
           split_dims.append(mesh_axis_names[s_i])
       elif isinstance(s_i, str):
-        assert s_i in mesh_axis_names
+        #assert s_i in mesh_axis_names
         split_dims.append(s_i)
       elif isinstance(s_i, (tuple, list)):
         split_dims.append(_parse_split_dims(s_i))
diff --git a/praxis/contrib/gpu/scripts_gpu/te_helper.py b/praxis/contrib/gpu/scripts_gpu/te_helper.py
new file mode 100644
index 0000000..d0afc1a
--- /dev/null
+++ b/praxis/contrib/gpu/scripts_gpu/te_helper.py
@@ -0,0 +1,321 @@
+import os
+
+from praxis import base_layer
+from praxis import pax_fiddle
+from praxis import pytypes
+from praxis import layers
+from praxis.layers.checkpoint_policy import AutodiffCheckpointType
+from praxis.layers import activations
+from praxis.layers import attentions, grouped_query_attention, multi_query_attention
+from praxis.layers import embedding_softmax
+from praxis.layers import normalizations
+
+try:
+    import transformer_engine.jax as te
+    import transformer_engine.jax.flax as te_flax
+    import transformer_engine.jax.praxis as te_praxis
+    _IS_TRANSFORMER_ENGINE_INSTALLED = True
+    import praxis.layers.repeats as praxis_repeat
+    # This is to make Repeat module correctly generate collections we need.
+    praxis_repeat.SCAN_VARIABLE_AXES.update({base_layer.NON_PAX_VAR_COLLECTION[1]: 0, # 1-idx = params_axes
+                                            te.fp8.FP8Helper.FP8_COLLECTION_NAME:0})
+    TE_PIPELINE_EXTRA_VMAP_VAR_AXES = {
+        base_layer.NON_PAX_VAR_COLLECTION[1]: 0, # 1-idx = params_axes
+        te.fp8.FP8Helper.FP8_COLLECTION_NAME:0
+    }
+
+    TE_PIPELINE_EXTRA_SCAN_VAR_BROADCAST = [te.fp8.FP8Helper.FP8_COLLECTION_NAME]
+
+    ENABLE_TE_SP = bool(int(os.environ.get('ENABLE_TE_SP', 0)))
+
+except ModuleNotFoundError as e:
+    _IS_TRANSFORMER_ENGINE_INSTALLED = False
+    TE_PIPELINE_EXTRA_VMAP_VAR_AXES = {}
+    TE_PIPELINE_EXTRA_SCAN_VAR_BROADCAST = []
+    ENABLE_TE_SP = False
+
+LayerTpl = pax_fiddle.Config[base_layer.BaseLayer]
+JTensor = pytypes.JTensor
+
+
+class TransformerEngineHelperBase:
+
+    @staticmethod
+    def get_fprop_caller_of_stack_transformer(fprop, deterministic):
+        raise NotImplementedError
+
+    @staticmethod
+    def set_layer_params_to_stack_transformer(stacked_transformer_obj, layer_p, layer_id):
+        raise NotImplementedError
+
+    @staticmethod
+    def get_input_bld(original_bld, batch_axes, mdl_axis):
+        # This is used to specify the sharding pattern of inputs to TransformerLayers.
+        raise NotImplementedError
+
+    @staticmethod
+    def get_bld_mapping_for_pipelined_transformer(xformer_layer_p):
+        raise NotImplementedError
+
+    @staticmethod
+    def check_checkpoint_policy(tpl):
+        raise NotImplementedError
+
+
+class TENotInstalledHelper(TransformerEngineHelperBase):
+
+    @staticmethod
+    def get_fprop_caller_of_stack_transformer(fprop, deterministic):
+        return fprop
+
+    @staticmethod
+    def set_layer_params_to_stack_transformer(stacked_transformer_obj, layer_p, layer_id):
+        layer_p.name = f'layer_{layer_id}'
+        layer_p.use_cross_attention = stacked_transformer_obj.use_cross_attention
+        layer_p.num_heads = stacked_transformer_obj.num_heads
+        layer_p.dim_per_head = stacked_transformer_obj.dim_per_head
+        layer_p.input_dims = stacked_transformer_obj.model_dims
+        layer_p.packed_input = stacked_transformer_obj.packed_input
+        layer_p.atten_dropout_prob = stacked_transformer_obj.atten_dropout_prob or stacked_transformer_obj.dropout_prob
+        layer_p.residual_dropout_prob = (
+            stacked_transformer_obj.residual_dropout_prob or stacked_transformer_obj.dropout_prob
+        )
+        layer_p.relu_dropout_prob = stacked_transformer_obj.relu_dropout_prob or stacked_transformer_obj.dropout_prob
+        layer_p.hidden_dims = stacked_transformer_obj.hidden_dims
+
+        if stacked_transformer_obj.residual_droppath_prob > 0.0:
+            layer_p.residual_droppath_prob = (
+                stacked_transformer_obj.residual_droppath_prob * layer_id / max(1, stacked_transformer_obj.num_layers)
+            )
+        return layer_p
+
+    @staticmethod
+    def get_input_bld(original_bld, *_):
+        return original_bld
+
+    @staticmethod
+    def get_bld_mapping_for_pipelined_transformer(xformer_layer_p):
+        return xformer_layer_p.tr_atten_tpl.activation_split_dims_mapping.bld
+
+    @staticmethod
+    def check_checkpoint_policy(_):
+        """Every checkpoint policy is valid without TE"""
+        pass
+
+
+class TEInstalledHelper(TransformerEngineHelperBase):
+
+    @staticmethod
+    def get_fprop_caller_of_stack_transformer(_, deterministic):
+        def _fprop(
+            transformer,
+            x_in,
+            paddings,
+            attention_mask,
+            cross_inputs,
+            cross_attention_mask,
+            segment_pos
+        ):
+            x_out = transformer(
+                inputs=x_in,
+                attention_mask=attention_mask,
+                encoded=cross_inputs,
+                encoder_decoder_mask=cross_attention_mask,
+                deterministic=deterministic)
+            return x_out
+        return _fprop
+
+
+    @staticmethod
+    def set_layer_params_to_stack_transformer(stacked_transformer_obj, _, layer_id):
+        te_transformer_tpl = pax_fiddle.Config(te_praxis.TransformerLayer,
+            name=f'layer_{layer_id}',
+            enable_relative_embedding=False,
+            enable_sequence_parallel=ENABLE_TE_SP,
+            transpose_batch_sequence=False
+        )
+
+        def update_ln_te_tpl(te_tpl, transformer_layer_tpl):
+            # TE requires all normalization are the same
+            assert transformer_layer_tpl.ln_tpl == transformer_layer_tpl.tr_fflayer_tpl.ln_tpl
+            ln_tpl = transformer_layer_tpl.ln_tpl
+            if issubclass(ln_tpl.cls, normalizations.LayerNorm):
+                te_tpl.layernorm_type = 'layernorm'
+                assert ln_tpl.use_scale
+                assert ln_tpl.use_bias
+            elif issubclass(ln_tpl.cls, normalizations.RmsNorm):
+                te_tpl.layernorm_type = 'rmsnorm'
+            else:
+                raise ValueError(f'Unsupported {ln_tpl.cls=}, LayerNorm, RmsNorm are supported.')
+            te_tpl.zero_centered_gamma = not ln_tpl.direct_scale
+            te_tpl.layernorm_epsilon = ln_tpl.epsilon
+            return te_tpl
+
+        def update_ff_te_tpl(te_tpl, ff_layer_tpl):
+            mlp_activations = ()
+            if ff_layer_tpl.use_gated_activation:
+               mlp_activations += ('linear',)
+
+            if issubclass(ff_layer_tpl.activation_tpl.cls, activations.Identity):
+                mlp_activations += ('linear',)
+            else:
+                mlp_activations += (ff_layer_tpl.activation_tpl.cls.__name__.lower(),)
+
+            te_tpl.mlp_activations = mlp_activations
+            return te_tpl
+
+        def update_attn_te_tpl(te_tpl, attn_tpl):
+            if issubclass(attn_tpl.cls, attentions.DotProductAttention):
+                # Check the DotProductAttention parameters are aligned to TE's attention
+                assert attn_tpl.internal_enable_query_scale or attn_tpl.scale_logits_by_head_dims
+                assert not (attn_tpl.internal_enable_query_scale and attn_tpl.scale_logits_by_head_dims)
+                assert not attn_tpl.internal_enable_per_dim_scale
+                assert not attn_tpl.scale_query_by_dim_per_head
+                assert not attn_tpl.dconv_qkv
+                assert not attn_tpl.internal_gshard_gaussian_init
+                assert attn_tpl.relative_bias_tpl is None
+                assert attn_tpl.attention_extra_logit is None
+                assert attn_tpl.ngrammer_tpl is None
+                te_tpl.enable_rotary_pos_emb = attn_tpl.use_rotary_position_emb
+                if issubclass(attn_tpl.rotary_position_emb_tpl.cls, embedding_softmax.RotaryPositionalEmbedding):
+                    te_tpl.rotary_pos_emb_group_method = 'alternate'
+            elif issubclass(attn_tpl.cls, grouped_query_attention.GroupedQueryAttention):
+                te_tpl.num_gqa_groups = attn_tpl.num_kv_heads
+                if attn_tpl.rope_min_max_timescales is not None:
+                    te_tpl.enable_rotary_pos_emb = True
+                    te_tpl.rotary_pos_emb_windows = attn_tpl.rope_min_max_timescales
+                assert attn_tpl.atten_temp == 1.
+            elif issubclass(attn_tpl.cls, multi_query_attention.MultiQueryDotProductAttention):
+                te_tpl.num_gqa_groups = attn_tpl.num_kv_heads
+                te_tpl.enable_rotary_pos_emb = attn_tpl.use_rotary_position_emb
+                if issubclass(attn_tpl.rotary_position_emb_tpl.cls, embedding_softmax.RotaryPositionalEmbedding):
+                    te_tpl.rotary_pos_emb_group_method = 'alternate'
+            else:
+                raise ValueError(f'Unsupported {attn_tpl.cls=}')
+            assert attn_tpl.atten_logit_cap <= 0., 'atten_logit_cap > 0. is not supported in TE'
+            te_tpl.scaled_query_init = False
+            te_tpl.scale_attn_logits = True
+            return te_tpl
+
+        transformer_layer_tpl = stacked_transformer_obj.transformer_layer_params_tpl
+        # Update TE normalization layer configs
+        te_transformer_tpl = update_ln_te_tpl(te_transformer_tpl, transformer_layer_tpl)
+        # Update TE feed forward layer configs
+        te_transformer_tpl = update_ff_te_tpl(te_transformer_tpl, transformer_layer_tpl.tr_fflayer_tpl)
+        # Update TE attention layer configs
+        te_transformer_tpl = update_attn_te_tpl(te_transformer_tpl, transformer_layer_tpl.tr_atten_tpl)
+        # TE currently only allow the bias config to be same between feed forward, qkv proj, out proj
+        assert (transformer_layer_tpl.tr_fflayer_tpl.has_bias ==
+            transformer_layer_tpl.tr_atten_tpl.use_bias), "TE only allows same bias settings."
+        te_transformer_tpl.use_bias = transformer_layer_tpl.tr_fflayer_tpl.has_bias
+        te_transformer_tpl.self_attn_mask_type = 'causal' \
+            if stacked_transformer_obj.mask_self_attention else 'padding'
+
+        te_transformer_tpl.logical_axes_rules = te_flax.extend_logical_axis_rules(tuple())
+        te_transformer_tpl.params_init = stacked_transformer_obj.params_init
+        te_transformer_tpl.dtype = stacked_transformer_obj.fprop_dtype
+        te_transformer_tpl.layer_type = te_praxis.TransformerLayerType.DECODER if stacked_transformer_obj.use_cross_attention \
+                        else te_praxis.TransformerLayerType.ENCODER
+        te_transformer_tpl.num_attention_heads = stacked_transformer_obj.num_heads
+        te_transformer_tpl.hidden_size = stacked_transformer_obj.model_dims
+        te_transformer_tpl.mlp_hidden_size = stacked_transformer_obj.hidden_dims
+        te_transformer_tpl.dropout_rng_name = base_layer.RANDOM
+        te_transformer_tpl.attention_dropout = stacked_transformer_obj.atten_dropout_prob or stacked_transformer_obj.dropout_prob
+        te_transformer_tpl.hidden_dropout = stacked_transformer_obj.residual_dropout_prob or stacked_transformer_obj.dropout_prob
+        te_transformer_tpl.intermediate_dropout = stacked_transformer_obj.relu_dropout_prob or stacked_transformer_obj.dropout_prob
+        if stacked_transformer_obj.residual_droppath_prob > 0.0:
+            te_transformer_tpl.drop_path = (
+                stacked_transformer_obj.residual_droppath_prob * layer_id / max(1, stacked_transformer_obj.num_layers)
+            )
+
+        assert stacked_transformer_obj.dim_per_head == stacked_transformer_obj.model_dims // stacked_transformer_obj.num_heads
+        assert stacked_transformer_obj.packed_input == False
+        assert len(stacked_transformer_obj.moe_layers) == 0
+        assert stacked_transformer_obj.ngrammer_tpls is None
+
+        return te_transformer_tpl
+
+    @staticmethod
+    def get_input_bld(_, batch_axes, mdl_axis):
+        if ENABLE_TE_SP:
+            return [batch_axes, mdl_axis, None]
+        return [batch_axes, None, None]
+
+    @staticmethod
+    def get_bld_mapping_for_pipelined_transformer(_):
+        rules = te_flax.extend_logical_axis_rules(tuple())
+        # rules [(batch_axis_name, ('replicat', 'data'))', ...)]
+        batch_mapping = rules[0][1]
+        hidden_tp_mapping = rules[4][1]
+        # [Batch, Seqlen, Hidden]
+        bld_mapping = [batch_mapping, None, hidden_tp_mapping]
+        return bld_mapping
+
+    @staticmethod
+    def check_checkpoint_policy(tpl):
+        """Some checkpoint policies are not compatible with TE fused attention"""
+        if issubclass(tpl.cls, layers.transformers.StackedTransformer):
+            remat = tpl.remat
+            attention_dropout = tpl.atten_dropout_prob or tpl.dropout_prob
+        elif issubclass(tpl.cls, layers.transformers.StackedTransformerRepeated):
+            if not issubclass(tpl.block.cls, layers.transformers.StackedTransformer):
+                return
+            remat = True  # Current StackedTransformerRepeat always enables remat
+            attention_dropout = tpl.block.atten_dropout_prob or tpl.block.dropout_prob
+        else:
+            raise ValueError(f'Unsupported class={tpl.cls}')
+
+        supported_checkpoint_policies = [
+            AutodiffCheckpointType.SAVE_CONTEXT,
+            AutodiffCheckpointType.SAVE_CONTEXT_AND_OUT_PROJ,
+            AutodiffCheckpointType.SAVE_DOT_FOR_MLPERF_200B,
+            AutodiffCheckpointType.SAVE_QUANTIZED,
+            AutodiffCheckpointType.SAVE_DOT_EXCEPT_LOGITS_FFN1,
+            AutodiffCheckpointType.SAVE_DOT_EXCEPT_LOGITS]
+        fused_attn_enabled = int(os.getenv("NVTE_FUSED_ATTN", "0"))
+        if remat and fused_attn_enabled and attention_dropout > 0.:
+            assert tpl.checkpoint_policy in supported_checkpoint_policies, \
+            "Fused attn in TE only permits policies that save 'context' tensors when dropout is " \
+            "enabled. This restriction is due to the maintenance of the dropout offset within TE, " \
+            "which is incompatible with the JAX remat. Consequently, it's necessary to bypass " \
+            "recomputation in the attention layer when fused attention is activated. The supported " \
+            f"checkpoint_policies are {supported_checkpoint_policies} but the provided " \
+            f"checkpoint_policy is '{tpl.checkpoint_policy}'."
+
+
+class TransformerEngineHelper(TransformerEngineHelperBase):
+
+    @staticmethod
+    def is_enabled_te():
+        enable_te = bool(int((os.environ.get("ENABLE_TE", False))))
+        return (_IS_TRANSFORMER_ENGINE_INSTALLED and enable_te)
+
+    @staticmethod
+    def get_helper():
+        if TransformerEngineHelper.is_enabled_te():
+            return TEInstalledHelper
+        return TENotInstalledHelper
+
+    @staticmethod
+    def get_fprop_caller_of_stack_transformer(fprop, deterministic):
+        return TransformerEngineHelper.get_helper().get_fprop_caller_of_stack_transformer(
+                    fprop, deterministic)
+
+    @staticmethod
+    def set_layer_params_to_stack_transformer(stacked_transformer_obj, layer_p, layer_id):
+        return TransformerEngineHelper.get_helper().set_layer_params_to_stack_transformer(
+                    stacked_transformer_obj, layer_p, layer_id)
+
+    @staticmethod
+    def get_input_bld(original_bld, batch_axes, mdl_axis):
+        return TransformerEngineHelper.get_helper().get_input_bld(
+                    original_bld, batch_axes, mdl_axis)
+
+    @staticmethod
+    def get_bld_mapping_for_pipelined_transformer(xformer_layer_p):
+        return TransformerEngineHelper.get_helper().get_bld_mapping_for_pipelined_transformer(
+                    xformer_layer_p)
+
+    @staticmethod
+    def check_checkpoint_policy(tpl):
+        return TransformerEngineHelper.get_helper().check_checkpoint_policy(tpl)
diff --git a/praxis/layers/attentions.py b/praxis/layers/attentions.py
index d853bbf..d036265 100644
--- a/praxis/layers/attentions.py
+++ b/praxis/layers/attentions.py
@@ -1182,6 +1182,7 @@ class DotProductAttention(base_layer.BaseLayer):
   decode_cache: bool = True
   attention_mask_summary: bool = False
   zero_fully_masked: bool = False
+  mha_mask_addition_pattern: bool = True
   qk_einsum_tpl: LayerTpl = template_field(base_ops.EinsumOp)
   pv_einsum_tpl: LayerTpl = template_field(base_ops.EinsumOp)
   per_dim_scale_tpl: LayerTpl = template_field(PerDimScale)
@@ -1541,7 +1542,11 @@ class DotProductAttention(base_layer.BaseLayer):
     # Attention softmax is always carried out in fp32.
     logits = logits.astype(jnp.float32)
     # Apply attention masking
-    padded_logits = py_utils.apply_mask_to_logits(logits, atten_mask)
+    if self.mha_mask_addition_pattern:
+      padded_logits = logits + atten_mask.astype(jnp.float32)
+    else:
+      padded_logits = py_utils.apply_mask_to_logits(logits, atten_mask)
+
     if self.attention_mask_summary:
       self.add_summary('attention_mask', atten_mask)
     if self.attention_extra_logit is None:
diff --git a/praxis/layers/pipeline.py b/praxis/layers/pipeline.py
index e3b2f7c..b31526e 100644
--- a/praxis/layers/pipeline.py
+++ b/praxis/layers/pipeline.py
@@ -28,6 +28,8 @@ from praxis import pax_fiddle
 from praxis import py_utils
 from praxis import pytypes
 from praxis.layers import checkpoint_policy
+from praxis.contrib.gpu.scripts_gpu.te_helper import TE_PIPELINE_EXTRA_VMAP_VAR_AXES
+from praxis.contrib.gpu.scripts_gpu.te_helper import TE_PIPELINE_EXTRA_SCAN_VAR_BROADCAST
 
 NestedMap = py_utils.NestedMap
 JTensor = pytypes.JTensor
@@ -414,6 +416,7 @@ class LayerwiseShardablePipelined(base_layer.BaseLayer):
             NON_TRAINABLE: 0,
             INTERMEDIATES: 0,
             HYPER_PARAMS: 0,
+            **TE_PIPELINE_EXTRA_VMAP_VAR_AXES
         },
         split_rngs={PARAMS: self.is_initializing(), RANDOM: True},
         metadata_params={
@@ -798,7 +801,7 @@ class LayerwiseShardablePipelined(base_layer.BaseLayer):
     #
     # Note that fprop should not use PARAMS rng because there is no var init.
     variable_carry = []
-    variable_broadcast = [PARAMS]
+    variable_broadcast = [PARAMS] + TE_PIPELINE_EXTRA_SCAN_VAR_BROADCAST
     if self.is_mutable_collection(NON_TRAINABLE):
       variable_carry.append(NON_TRAINABLE)
     else:
@@ -821,7 +824,7 @@ class LayerwiseShardablePipelined(base_layer.BaseLayer):
     if bf16_vars_to_convert is not None:
       scan_fn = nn.map_variables(
           scan_fn,
-          mapped_collections=[PARAMS],
+          mapped_collections=[PARAMS, 'fp8_meta_collection'],
           mutable=True,
           trans_in_fn=_get_to_f32_converter(bf16_vars_to_convert),
           trans_out_fn=_get_to_bf16_converter(bf16_vars_to_convert),
diff --git a/praxis/layers/transformer_models.py b/praxis/layers/transformer_models.py
index 9b87c9f..ce919c8 100644
--- a/praxis/layers/transformer_models.py
+++ b/praxis/layers/transformer_models.py
@@ -33,6 +33,7 @@ from praxis.layers import embedding_softmax
 from praxis.layers import multi_query_attention
 from praxis.layers import normalizations
 from praxis.layers import transformers
+from praxis.contrib.gpu.scripts_gpu.te_helper import TransformerEngineHelper
 
 NestedMap = py_utils.NestedMap
 JTensor = pytypes.JTensor
@@ -126,6 +127,99 @@ def _set_embedding_softmax_sharding_params_for_transformers(
   return embedding_softmax_p
 
 
+def _set_stacked_transformer_sharding_with_expert_parallelism(
+    stacked_transformer_p,
+    *,
+    w_df,
+    w_dnh,
+    w_emh,
+    a_bld,
+    a_blf,
+    a_blnh,
+    a_egch,
+    a_egcm,
+    a_blh=None,
+    a_bd=None,
+    a_bf=None,
+):
+  """Set sharding params for the stacked transformer layer."""
+  stacked_p = stacked_transformer_p
+  if fdl.get_callable(stacked_p) == transformers.PipelinedTransformer:
+    stacked_p = stacked_p.pipeline_stage
+  if issubclass(
+      fdl.get_callable(stacked_p), transformers.StackedTransformerRepeated
+  ):
+    stacked_p = stacked_p.block
+  transformer_p = stacked_p.transformer_layer_params_tpl
+  if isinstance(transformer_p, Sequence):
+    transformer_p_lst = transformer_p
+  else:
+    transformer_p_lst = [transformer_p]
+  for t_p in transformer_p_lst:
+    for atten_p in (t_p.tr_atten_tpl, t_p.cross_atten_tpl):
+      if atten_p is None:
+        continue
+      atten_wp = atten_p.weight_split_dims_mapping
+      atten_wp.proj = w_dnh
+      w_n_sharding = None if w_dnh is None else w_dnh[1]
+      atten_wp.dconv = [w_n_sharding, None]
+      atten_ap = atten_p.activation_split_dims_mapping
+      atten_ap.blnh = a_blnh
+      atten_ap.bld = a_bld
+      if hasattr(atten_ap, 'bd'):
+        atten_ap.bd = a_bd
+      if issubclass(
+          fdl.get_callable(atten_p),
+          multi_query_attention.MultiQueryDotProductAttention,
+      ):
+        atten_wp.proj_headless = [w_dnh[0], w_dnh[2]]
+        if a_blh is None:
+          atten_ap.blh = [a_blnh[0], a_blnh[1], a_blnh[3]]
+        else:
+          atten_ap.blh = a_blh
+
+    ff_p = t_p.tr_fflayer_tpl
+    ff_wp = ff_p.weight_split_dims_mapping
+    ff_wp.ffn0 = w_df
+    if w_df is None:
+      ff_wp.ffn1 = None
+    else:
+      ff_wp.ffn1 = [w_df[1], w_df[0]]
+    ff_ap = ff_p.activation_split_dims_mapping
+    ff_ap.ffn0 = a_blf
+    if hasattr(ff_ap, 'ffn0_extend_step'):
+      ff_ap.ffn0_extend_step = a_bf
+    ff_ap.ffn1 = a_bld
+    if hasattr(ff_ap, 'ffn1_extend_step'):
+      ff_ap.ffn1_extend_step = a_bd
+
+  if stacked_p.moe_layer_tpl is not None:
+    # Set Moe layer sharding hparams.
+    moe_p = stacked_p.moe_layer_tpl
+    moe_wp = moe_p.weight_split_dims_mapping
+    moe_wp.me = [None, None]  # Replicated.
+    moe_wp.emh = w_emh
+    w_e_sharding = None if w_emh is None else w_emh[0]
+    w_m_sharding = None if w_emh is None else w_emh[1]
+    w_h_sharding = None if w_emh is None else w_emh[2]
+    moe_wp.ehm = [w_e_sharding, w_h_sharding, w_m_sharding]
+    # Activations
+    a_e_sharding = None if a_egch is None else ('data', 'data_expert')
+    moe_ap = moe_p.activation_split_dims_mapping
+    moe_ap.gs = [a_e_sharding, None]
+    # dispatch and combine tensors
+    moe_ap.gsec = [a_e_sharding, None, None, None]
+    moe_ap.gecs = [a_e_sharding, None, None, None]
+    moe_ap.gec = [a_e_sharding, None, None]
+    moe_ap.egch = a_egch
+    a_e_sharding = None if a_egcm is None else ('data', 'data_expert')
+    a_m_sharding = None if a_egcm is None else a_egcm[3]
+    moe_ap.gsm = [a_e_sharding, None, a_m_sharding]
+    moe_ap.egcm = a_egcm
+    moe_ap.gecm = [a_egcm[1], a_egcm[0], a_egcm[2], a_egcm[3]]
+  return stacked_transformer_p
+
+
 def _set_stacked_transformer_sharding(
     stacked_transformer_p,
     *,
@@ -291,6 +385,108 @@ class TransformerLm(base_layer.BaseLayer):
   record_activations_in_xent_output: bool = False
   entropy_loss_weight: float | None = None
 
+  @classmethod
+  def set_sharding_params_with_expert_parallelism(
+      cls,
+      lm_p,
+      *,
+      replica_axis,
+      data_axis,
+      data_expert_axis,
+      mdl_axis,
+      ici_mesh_shape,
+      dcn_mesh_shape=None,
+      batch_axes=None,
+      mesh_axis_names,
+      seq_axis=None,
+      training_optimized,
+  ):
+    """Set Canonical sharding params.
+
+    Args:
+      lm_p: A params of this class.
+      replica_axis: A string or int of the model replica axis name.
+      data_axis: A string or int of the data axis name.
+      data_expert_axis: A string or int of the data/expert axis name.
+      mdl_axis: A string or int of the mdl axis name.
+      ici_mesh_shape: Shape of logical mesh for a slice.
+      dcn_mesh_shape: Shape of logical mesh between slices.
+      batch_axes: A tuple with all axes used for splitting activations along
+        batch dimension. Defaults to (replica_axis, data_axis).
+      mesh_axis_names: A list of length len(shape). Each element of the list is
+        the name of the corresponding device axis.
+      training_optimized: A bool indicating whether sharding is optimized for
+        training by saving activation memory between forward and backward
+        passes.
+
+    Returns:
+      Params with sharding annotations added.
+    """
+    # In the following, model weights are layed out on the [data_axis, mdl_axis]
+    # 2d mesh. Model weights are always replicated over the replica_axis mesh
+    # axis.
+    #
+    # The batch axis of the activations are always sharded over the combination
+    # of (replica_axis, data_axis).
+    if batch_axes is None:
+      batch_axes = (replica_axis, data_axis, data_expert_axis)
+    bld = (
+        [batch_axes, seq_axis, mdl_axis]
+        if training_optimized
+        else [batch_axes, None, None]
+    )
+    egcm = (
+        [data_expert_axis, data_axis, None, mdl_axis]
+        if training_optimized
+        else [batch_axes, None, None, None]
+    )
+
+    if seq_axis is None:
+      w_data_axes = (data_axis, data_expert_axis)
+    else:
+      w_data_axes = (data_axis, data_expert_axis, seq_axis)
+
+    # w_df: sharding for weight of ffn0, shape (d, f). ff1 weights will be
+    # inferred from it.
+    w_df = [w_data_axes, mdl_axis]
+    # w_dnh: Sharding of qkv projection weights, shape (d, num_heads,
+    # per_head_size)
+    w_dnh = [w_data_axes, mdl_axis, None]
+    # w_emh: sharding for first MoE FFN weight, shape (e, m, h). The second MoE
+    # ffn weight will be inferred from it.
+    w_emh = [data_expert_axis, data_axis, mdl_axis]
+    # w_vd: sharding of the embedding weight of (vocab_size, d).
+    w_vd = [mdl_axis, w_data_axes]
+    # a_bld: sharding of output of ffn/attention, shape (b, l, d).
+    a_bld = bld
+    # a_blf: sharding of output of ffn0, shape (b, l, f).
+    a_blf = [batch_axes, seq_axis, mdl_axis]
+    # a_blnh: sharding of the attention activation of shape (b, l, num_heads,
+    # per_head_size).
+    a_blnh = [batch_axes, seq_axis, mdl_axis, None]
+    # a_blv: sharding of the logits activation of shape (b, l, vocab_size).
+    a_blv = [batch_axes, seq_axis, mdl_axis]
+    # a_egch: sharding of the output of first MoE FFN, shape (e, g, c, h).
+    a_egch = [data_expert_axis, data_axis, None, mdl_axis]
+    # a_egcm: sharding of the output of second MoE FFN, shape (e, g, c, m).
+    a_egcm = egcm
+    return cls.set_custom_sharding_params(
+        lm_p,
+        ici_mesh_shape=ici_mesh_shape,
+        dcn_mesh_shape=dcn_mesh_shape,
+        mesh_axis_names=mesh_axis_names,
+        w_df=w_df,
+        w_dnh=w_dnh,
+        w_emh=w_emh,
+        w_vd=w_vd,
+        a_bld=a_bld,
+        a_blf=a_blf,
+        a_blnh=a_blnh,
+        a_blv=a_blv,
+        a_egch=a_egch,
+        a_egcm=a_egcm,
+    )
+
   @classmethod
   def set_sharding_params_v1(
       cls,
@@ -340,6 +536,8 @@ class TransformerLm(base_layer.BaseLayer):
         if training_optimized
         else [batch_axes, None, None]
     )
+    bld = TransformerEngineHelper.get_input_bld(bld, batch_axes, mdl_axis)
+
     egcm = (
         [data_axis, None, None, mdl_axis]
         if training_optimized
@@ -449,6 +647,9 @@ class TransformerLm(base_layer.BaseLayer):
     lm_p.ici_mesh_shape = ici_mesh_shape
     lm_p.dcn_mesh_shape = dcn_mesh_shape
     lm_p.mesh_axis_names = mesh_axis_names
+    use_expert_parallem = (
+        mesh_axis_names is not None and 'data_expert' in mesh_axis_names
+    )
     pos_emb_w_ld = w_df
     if (
         lm_p.position_emb_tpl is not None
@@ -477,7 +678,22 @@ class TransformerLm(base_layer.BaseLayer):
         **mesh_kwargs,
     )
 
-    def _set_transformer_sharding(transformer_tpl):
+    def _set_transformer_sharding(transformer_tpl, use_expert_parallem=False):
+      if use_expert_parallem:
+        return _set_stacked_transformer_sharding_with_expert_parallelism(
+            transformer_tpl,
+            w_df=w_df,
+            w_dnh=w_dnh,
+            w_emh=w_emh,
+            a_bld=a_bld,
+            a_blf=a_blf,
+            a_blh=a_blh,
+            a_blnh=a_blnh,
+            a_bd=a_bd,
+            a_bf=a_bf,
+            a_egch=a_egch,
+            a_egcm=a_egcm,
+        )
       return _set_stacked_transformer_sharding(
           transformer_tpl,
           w_df=w_df,
@@ -495,11 +711,11 @@ class TransformerLm(base_layer.BaseLayer):
 
     if lm_p.stacked_transformer_tpl.cls == transformers.PipelinedTransformer:
       lm_p.stacked_transformer_tpl.pipeline_stage = _set_transformer_sharding(
-          lm_p.stacked_transformer_tpl.pipeline_stage
+          lm_p.stacked_transformer_tpl.pipeline_stage, use_expert_parallem
       )
     else:
       lm_p.stacked_transformer_tpl = _set_transformer_sharding(
-          lm_p.stacked_transformer_tpl
+          lm_p.stacked_transformer_tpl, use_expert_parallem
       )
 
     if lm_p.separate_embedding_tpl is not None:
@@ -1804,26 +2020,26 @@ class TransformerEncoderDecoder(base_layer.BaseLayer):
     xent_output.total_loss = xent_output.avg_xent + xent_output.aux_loss
     return xent_output
 
-  def __call__(
+  def fprop_with_encoder_output(
       self,
-      inputs: JTensor,
+      encoder_output: JTensor,
       input_paddings: JTensor,
       targets: JTensor,
       target_paddings: JTensor,
       labels: NestedMap | None = None,
       input_segment_ids: JTensor | None = None,
       input_segment_pos: JTensor | None = None,
-      input_segment_mask: JTensor | None = None,
       target_segment_ids: JTensor | None = None,
       target_segment_pos: JTensor | None = None,
       target_segment_mask: JTensor | None = None,
       cross_segment_mask: JTensor | None = None,
       start_time_step: int = 0,
   ) -> NestedMap:
-    """Computes xent loss given the sequence model inputs.
+    """Computes xent loss given the encoder output.
 
     Args:
-      inputs: Input ids. An int32 JTensor of shape [B, S].
+      encoder_output: A JTensor of shape [B, S, D]. It's the output of the
+        encoder for the input.
       input_paddings: A 0/1 JTensor of shape [B, S] with 1 denoting padding
         corresponding to the input sequence.
       targets: Target ids. An int32 JTensor of shape [B, T].
@@ -1838,8 +2054,6 @@ class TransformerEncoderDecoder(base_layer.BaseLayer):
         token belongs to.
       input_segment_pos: A JTensor of shape [B, S]. The position of each input
         token within a segment.
-      input_segment_mask: A JTensor or shape [B, 1, S, S]. The segment mask for
-        packed input tokens.
       target_segment_ids: A JTensor of shape [B,T]. The segment that each target
         token belongs to.
       target_segment_pos: A JTensor of shape [B, T]. The position of each target
@@ -1859,14 +2073,6 @@ class TransformerEncoderDecoder(base_layer.BaseLayer):
     """
     batch, target_seq_length = targets.shape[:2]
 
-    encoder_output = self.encode(
-        inputs,
-        input_paddings,
-        input_segment_ids,
-        input_segment_pos,
-        input_segment_mask,
-    )
-
     if self.decoder_embedding_tpl is not None:
       # Targets have separate embedding params.
       target_emb = self.decoder_embedding_lookup.emb_lookup(targets)
@@ -1934,6 +2140,83 @@ class TransformerEncoderDecoder(base_layer.BaseLayer):
 
     return self.compute_loss(output, labels)
 
+  def __call__(
+      self,
+      inputs: JTensor,
+      input_paddings: JTensor,
+      targets: JTensor,
+      target_paddings: JTensor,
+      labels: NestedMap | None = None,
+      input_segment_ids: JTensor | None = None,
+      input_segment_pos: JTensor | None = None,
+      input_segment_mask: JTensor | None = None,
+      target_segment_ids: JTensor | None = None,
+      target_segment_pos: JTensor | None = None,
+      target_segment_mask: JTensor | None = None,
+      cross_segment_mask: JTensor | None = None,
+      start_time_step: int = 0,
+  ) -> NestedMap:
+    """Computes xent loss given the sequence model inputs.
+
+    Args:
+      inputs: Input ids. An int32 JTensor of shape [B, S].
+      input_paddings: A 0/1 JTensor of shape [B, S] with 1 denoting padding
+        corresponding to the input sequence.
+      targets: Target ids. An int32 JTensor of shape [B, T].
+      target_paddings: A 0/1 JTensor of shape [B, T] with 1 denoting padding
+        corresponding to the target sequence.
+      labels: A `.NestedMap` containing the following fields: class_weights, a
+        JTensor with shape [batch, seqlen] containing weights for each target
+        word. class_ids, a JTensor with shape [B, T] of int32 dtype containing
+        the target class labels. class_probabilities, a JTensor with shape [B,
+        T, V] of float values indicating class-membership probabilities.
+      input_segment_ids: A JTensor of shape [B,S]. The segment that each input
+        token belongs to.
+      input_segment_pos: A JTensor of shape [B, S]. The position of each input
+        token within a segment.
+      input_segment_mask: A JTensor or shape [B, 1, S, S]. The segment mask for
+        packed input tokens.
+      target_segment_ids: A JTensor of shape [B,T]. The segment that each target
+        token belongs to.
+      target_segment_pos: A JTensor of shape [B, T]. The position of each target
+        token within a segment.
+      target_segment_mask: A JTensor or shape [B, 1, T, T]. The segment mask for
+        packed target tokens.
+      cross_segment_mask: A JTensor or shape [B, 1, T, S]. The encoder-decoder
+        segment mask.
+      start_time_step: Decode extend_step start time step. When decoding after
+        prefix, start_time_step will be prefix_len - 1.
+
+    Returns:
+      Returns xent_output, where
+      `xent_output` is a `.NestedMap` as defined by `SoftmaxLayer`'s return. In
+      addition, per_sequence_xent is added which equal to the sum of xent loss
+      for tokens in a sequence.
+    """
+    batch, target_seq_length = targets.shape[:2]
+
+    encoder_output = self.encode(
+        inputs,
+        input_paddings,
+        input_segment_ids,
+        input_segment_pos,
+        input_segment_mask,
+    )
+    return self.fprop_with_encoder_output(
+        encoder_output=encoder_output,
+        input_paddings=input_paddings,
+        targets=targets,
+        target_paddings=target_paddings,
+        labels=labels,
+        input_segment_ids=input_segment_ids,
+        input_segment_pos=input_segment_pos,
+        target_segment_ids=target_segment_ids,
+        target_segment_pos=target_segment_pos,
+        target_segment_mask=target_segment_mask,
+        cross_segment_mask=cross_segment_mask,
+        start_time_step=start_time_step,
+    )
+
   def transform_decode_state(
       self, transform_fn: base_layer.DecodeStateTransformFn
   ) -> None:
diff --git a/praxis/layers/transformers.py b/praxis/layers/transformers.py
index e593112..0cbde48 100644
--- a/praxis/layers/transformers.py
+++ b/praxis/layers/transformers.py
@@ -39,6 +39,7 @@ from praxis.layers import pipeline
 from praxis.layers import repeats
 from praxis.layers import stats
 from praxis.layers import stochastics
+from praxis.contrib.gpu.scripts_gpu.te_helper import TransformerEngineHelper
 
 NestedMap = py_utils.NestedMap
 WeightInit = base_layer.WeightInit
@@ -1366,12 +1367,15 @@ class Transformer(base_layer.BaseLayer):
       atten_probs: A NestedMap with keys `self_atten` <float>[B, N, T, T].
     """
 
+    """
     inputs_stats = stats.compute_stats(inputs, jnp.expand_dims(paddings, -1))
     self.add_summary('xformer_input_mean', inputs_stats.mean_v, verbosity=3)
     self.add_summary('xformer_input_std', inputs_stats.std_v, verbosity=3)
     self.add_summary('xformer_input_abs_max', inputs_stats.max_v, verbosity=3)
 
     self.add_summary('attention_input_rms', _rms(inputs), verbosity=4)
+    """
+    
     if self.norm_policy == 'primer_hybrid':
       inputs_normalized = self.pre_layer_norm(inputs)
     elif self.norm_policy == 'pre':
@@ -1382,7 +1386,7 @@ class Transformer(base_layer.BaseLayer):
     # Compute self-attention, key/value vectors are the input itself
     atten_output, self_atten_probs = self.self_attention(
         inputs_normalized,
-        inputs_normalized,
+         inputs_normalized,
         inputs_normalized,
         atten_mask=attention_mask,
         query_segment_pos=segment_pos,
@@ -1717,23 +1721,10 @@ class StackedTransformer(base_layer.BaseLayer):
         p_i = self._clone_layer_params(self.transformer_layer_params_tpl[ii])
       else:
         p_i = self._clone_layer_params(self.transformer_layer_params_tpl)
-      p_i.name = f'layer_{i}'
-      p_i.use_cross_attention = self.use_cross_attention
-      p_i.num_heads = self.num_heads
-      p_i.dim_per_head = self.dim_per_head
-      p_i.input_dims = self.model_dims
-      p_i.packed_input = self.packed_input
-      p_i.atten_dropout_prob = self.atten_dropout_prob or self.dropout_prob
-      p_i.residual_dropout_prob = (
-          self.residual_dropout_prob or self.dropout_prob
-      )
-      p_i.relu_dropout_prob = self.relu_dropout_prob or self.dropout_prob
-      p_i.hidden_dims = self.hidden_dims
 
-      if self.residual_droppath_prob > 0.0:
-        p_i.residual_droppath_prob = (
-            self.residual_droppath_prob * i / max(1, self.num_layers)
-        )
+      TransformerEngineHelper.check_checkpoint_policy(self._to_fdl_config())
+
+      p_i = TransformerEngineHelper.set_layer_params_to_stack_transformer(self, p_i, i)
 
       if self.moe_layers and i in self.moe_layers:
         assert self.num_experts > 0
@@ -1852,6 +1843,8 @@ class StackedTransformer(base_layer.BaseLayer):
       )
       return x_out
 
+    _fprop = TransformerEngineHelper.get_fprop_caller_of_stack_transformer(_fprop, self.do_eval)
+
     fprop = _fprop
     if self.remat:
       fprop = nn.remat(
@@ -2044,6 +2037,8 @@ class StackedTransformerRepeated(base_layer.BaseLayer):
   def setup(self) -> None:
     wp = self.weight_split_dims_mapping
 
+    TransformerEngineHelper.check_checkpoint_policy(self._to_fdl_config())
+
     repeat_l_params = pax_fiddle.Config(
         repeats.Repeat,
         sub_tpl=self.block,
@@ -2321,7 +2316,7 @@ class PipelinedTransformer(base_layer.BaseLayer):
     else:
       assert self.pipeline_stage.cls == StackedTransformerRepeated
       xformer_layer_p = self.pipeline_stage.block.transformer_layer_params_tpl
-    bld_mapping = xformer_layer_p.tr_atten_tpl.activation_split_dims_mapping.bld
+    bld_mapping = TransformerEngineHelper.get_bld_mapping_for_pipelined_transformer(xformer_layer_p)
     if not self.stream_io:
       # Annotate the inputs before the pipeline to prevent unexpected
       # propagation from earlier layers.
diff --git a/praxis/py_utils.py b/praxis/py_utils.py
index a4f9c94..63eae4c 100644
--- a/praxis/py_utils.py
+++ b/praxis/py_utils.py
@@ -517,6 +517,7 @@ def maybe_pad_uneven_sharding(
   """Pads xs to make them evenly shardable, if needed."""
 
   def _maybe_pad(x, pspec, shape):
+    return x
     if is_optax_masked_node(x):
       return x
     paddings = get_uneven_sharding_paddings(pspec, shape, mesh_shape,
